### ISSUE TITLE
Tensorflow 2.3 doesn't log batch metrics and learning rate in Tensorboard

### ISSUE THREAD

**Issue by anth2o** [tensorflow/tensorflow#43621]
<em>Please make sure that this is a bug. As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Mac OS 10.14 or Ubuntu 18.04.3 (bug happens on both platforms)
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: Nope
- TensorFlow installed from (source or binary): pip
- TensorFlow version (use command below): 2.3 (bug) / 2.2 (good behavior)
- Python version: 3.6.8
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version: None
- GPU model and memory: None

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with:
1. TF 1.0: `python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"`
2. TF 2.0: `python -c "import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**

I used to make custom training loops to fit my models, with a tensorboard callback to log metrics, losses and learning rate using the following command:

```python
    callbacks.append(tf.keras.callbacks.TensorBoard(
        tensorboard_dir,
        profile_batch=0,
        update_freq=2, # write in tensorboard metrics and losses every 2 batches
        write_graph=False)
    )
```

Now, after upgrading tensorflow from 2.2 to 2.3, the same code doesn't log in tensorboard the metrics and losses one every 2 batches, and also the learning rate for each epoch (that I manage either by using `tf.keras.callbacks.LearningRateScheduler`
 or `tf.keras.callbacks.ReduceLROnPlateau`).

I shared a link below to a Colab with a minimal example, and the tensorboard for a model trained in 2.2 and in 2.3.

**Describe the expected behavior**

Same behavior in tf 2.3 as in tf 2.2.

**Standalone code to reproduce the issue**

https://colab.research.google.com/drive/16G5b0wfcG0SuNWrbQbv1pQfPbk6Ix9qU#scrollTo=7tuJE8iW_-NY

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.

**Comment by EmGarr** [comment]
It seems the behavior has been introduced by this commit: https://github.com/tensorflow/tensorflow/commit/69565ec4003902794bc94e10ba5fe9469a0b3ae4

The `self._log_metrics(logs, prefix='batch_', step=train_batches)` has been removed. For epoch log t `def _log_epoch_metrics(self, epoch, logs)` has been introduced. It seems that there should have an automatic update with

```python
if self.update_freq == 'epoch':
      should_record = False
      writer = None
    else:
      should_record = lambda: math_ops.equal(step % self.update_freq, 0)

    summary_state.is_recording = should_record
    summary_state.writer = writer
    # TODO(b/151339474): Fix deadlock when not using .value() here.
    summary_ops_v2.set_step(step.value())
```
However, I really don't know how it works.

@omalleyt12 do you have any information about this issue ?
Thanks

**Comment by amahendrakar** [comment]
Was able to reproduce the issue with TF v2.3 and TF-nightly, please find the gist of it [here](https://colab.research.google.com/gist/amahendrakar/5151b55dfb1341da1329f858c3c49a8d/43621-tf-nightly.ipynb). Thanks!

**Comment by omalleyt12** [comment]
> I used to make custom training loops to fit my models, with a tensorboard callback to log metrics, losses and learning rate using the following command:

Ah I understand now, the `TensorBoard` callback is designed for use with `Model.fit`. It still logs batch-level metrics when used in `Model.fit`. For a custom training loop, you should use lower-level `tf.summary` APIs to log batch-level summaries. The change from 2.2 to 2.3 made the batch-level summaries part of the `Model.train_function` rather than something that the `TensorBoard` callback creates itself. This resulted in a 2x improvement in speed for many small models in `Model.fit`, but it does have the side effect that calling `TensorBoard.on_train_batch_end(my_batch, my_metrics)` in a custom training loop will no longer log batch-level metrics
